slimt/Annotation.cc

```diff
@@ -74,4 +79,99 @@ void AnnotatedText::record_existing_sentence(
   annotation.token_begin_.push_back(text.size());
 }
 
+void AnnotatedText::to(Encoding encoding) {
+  if (encoding == encoding_) {
+    return;
+  }
+
+  if (encoding == Encoding::Byte && encoding_ == Encoding::UTF8) {
+    WordIterator current(*this);
+
+    std::vector<Range> words;
+
+    size_t byte_idx = 0;
+    Range byte{.begin = byte_idx, .end = 0};
+
+    const char *marker = text.data();
+    for (; current.has_next(); ++current) {
+      byte.begin = byte_idx;
+
+      for (size_t idx = (*current).begin; idx != (*current).end; idx++) {
+        ++marker;
+        ++byte_idx;
+
+        if ((*marker & 0xc0) == 0x80) {
+          ++byte_idx;
+          ++marker;
+        }
+      }
+      byte.end = byte_idx;
+      words.push_back(byte);
+    }
+
+    annotation.update(words);
+    encoding_ = Encoding::Byte;
+  } else if (encoding == Encoding::UTF8 && encoding_ == Encoding::Byte) {
+    WordIterator current(*this);
+
+    std::vector<Range> words;
+
+    size_t utf8_idx = 0;
+    size_t byte_idx = 0;
+    Range utf8{
+        //
+        .begin = utf8_idx,  //
+        .end = 0            //
+    };
+    // This loops run on the entire string.
+    //
+    // We have indices into two views of the same string. One is bytes, the
+    // other is utf8 encoded. We want to convert what is bytes to utf8.
+    //
+    // For this, we traverse through the characters, checking for unicode
+    // encoding and associated adjustments to the utf8_idx, keeping
+    // correspondences with the byte_idx;
+    for (const char c : text) {
+      // current = [begin, end)
+      if ((c & 0xc0) != 0x80)  // if is not utf-8 continuation character
+        ++utf8_idx;
```

Comment:
This works only for European Languages and Arabic, for full UTF-8 index mapping we have to also account for 3-byte or 4-byte unicode sequences in addition to the 2 byte and 1 byte sequence accounted for here.

https://chat.openai.com/share/e070b0e8-ad39-4a9f-8bf3-29b8a5236392

Reply:
We can use [emojis](https://news.ycombinator.com/item?id=15306020) to check multi-byte sequences, I think. I remember you put in european accented characters, can you test emojis as well within the higher level Python script and gold truth we have?

Reply:
I traced the missing stackoverflow answer - https://stackoverflow.com/a/3586973/4565794
This is my bad, wrong assumptions. The code skips every character that's a continuation so indexing is correct. 

Apologies. 

We may remove the function (or correct it and use it), but keep the tests so we know they map correctly.

Reply:
https://github.com/jerinphilip/slimt/pull/38/commits/7a9c418a5965d2c8c605656c8e31966d255eef2c